lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+		let mut act_three_builder = self.act_three_builder;
+		let bytes_read = act_three_builder.fill(input);
+
+		// In the event of a partial fill, stay in the same state and wait for more data
+		if !act_three_builder.is_finished() {
+			assert_eq!(bytes_read, input.len());
+			return Ok((
+				None,
+				HandshakeState::ResponderAwaitingActThree(Self {
+					hash: self.hash,
+					responder_ephemeral_private_key: self.responder_ephemeral_private_key,
+					chaining_key: self.chaining_key,
+					temporary_key: self.temporary_key,
+					act_three_builder
+				})
+			));
+		}
+
+		let hash = self.hash;
+		let temporary_key = self.temporary_key;
+		let responder_ephemeral_private_key = self.responder_ephemeral_private_key;
+		let chaining_key = self.chaining_key;
+
+		// 1. Read exactly 66 bytes from the network buffer
+		let act_three_bytes = Act::from(act_three_builder);
+		assert_eq!(act_three_bytes.len(), ACT_THREE_LENGTH);
+
+		// 2. Parse the read message (m) into v, c, and t
+		let version = act_three_bytes[0];
+		let tagged_encrypted_pubkey = &act_three_bytes[1..50];
+		let chacha_tag = &act_three_bytes[50..];
+
+		// 3. If v is an unrecognized handshake version, then the responder MUST abort the connection attempt.
+		if version != 0 {
+			// this should not crash the process, hence no panic
+			return Err("unexpected version".to_string());
+		}
+
+		// 4. rs = decryptWithAD(temp_k2, 1, h, c)
+		let mut remote_pubkey = [0; 33];
+		chacha::decrypt(&temporary_key, 1, &hash, &tagged_encrypted_pubkey, &mut remote_pubkey)?;
+		let initiator_pubkey = if let Ok(public_key) = PublicKey::from_slice(&remote_pubkey) {
+			public_key
+		} else {
+			return Err("invalid remote public key".to_string());
+		};
+
+		// 5. h = SHA-256(h || c)
+		let hash = concat_then_sha256!(hash, tagged_encrypted_pubkey);
+
+		// 6. se = ECDH(e.priv, rs)
+		let ecdh = ecdh(&responder_ephemeral_private_key, &initiator_pubkey);
+
+		// 7. ck, temp_k3 = HKDF(ck, se)
+		let (chaining_key, temporary_key) = hkdf5869rfc::derive(&chaining_key, &ecdh);
+
+		// 8. p = decryptWithAD(temp_k3, 0, h, t)
+		chacha::decrypt(&temporary_key, 0, &hash, &chacha_tag, &mut [0; 0])?;
+
+		// 9. rk, sk = HKDF(ck, zero)
+		let (receiving_key, sending_key) = hkdf5869rfc::derive(&chaining_key, &[0; 0]);
+
+		// 10. rn = 0, sn = 0
+		// - done by Conduit
+		let mut conduit = Conduit::new(sending_key, receiving_key, chaining_key);
+
+		// Any remaining data in the read buffer would be encrypted, so transfer ownership
+		// to the Conduit for future use.
+		conduit.read(&input[bytes_read..]);
+
+		Ok((
+			None,
+			HandshakeState::Complete(Some((conduit, initiator_pubkey)))
+		))
+	}
+}
+
+// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#handshake-state-initialization
+fn handshake_state_initialization(responder_static_public_key: &PublicKey) -> (Sha256, Sha256) {
```

Comment:
this method should probably be called `initialize_handshake_state`

Reply:
squashed into `review: Non-controversial renames, nits, small fixups` in the (1/4) PR